meshnets/data_processing/data_statistics.py

```diff
@@ -0,0 +1,78 @@
+"""File with functions to compute data statistics"""
+import numpy as np
+
+
+class OnlineStatistics:
+    """Computes mean and standard deviation from streaming data
+
+    Args:
+      axis: Axis over which to compute. Default is None, which means
+        that the statistics are computed over the flattened array.
+
+    Example:
+
+    >>> dataset = [[1, 2], [3, 4]]
+    >>> stats = OnlineStatistics()
+    >>> for data in dataset:
+    ...     stats.update(data)
+    >>> stats.get_mean()
+    array([[2.5]])
+    >>> stats.get_std()
+    array([[1.11803399]])
+    >>> stats = OnlineStatistics(axis=0)
+    >>> for data in dataset:
+    ...     stats.update(data)
+    >>> stats.get_mean()
+    array([[2., 3.]])
+    >>> stats.get_std()
+    array([[1., 1.]])
+    >>> stats = OnlineStatistics(axis=1)
+    >>> for data in dataset:
+    ...     stats.update(data)
+    >>> stats.get_mean()
+    array([[1.5], [3.5]])
+    >>> stats.get_std()
+    array([[0.5], [0.5]])
+
+    """
+
+    def __init__(self, axis=None):
+        self.axis = axis
+
+        self.n = None
+        self.cumulative_sum = None
+
+        self.cumulative_sum_squared = None
+
+    def update(self, example):
+        data_point = np.expand_dims(example, 0)
+        sum_element = np.sum(data_point, axis=self.axis, keepdims=True)
+        sum_element_squared = np.sum(data_point**2,
+                                     axis=self.axis,
+                                     keepdims=True)
+        if self.cumulative_sum is None:
+            self.cumulative_sum = sum_element
+            self.cumulative_sum_squared = sum_element_squared
+        else:
+            self.cumulative_sum = np.sum(np.concatenate(
```

Comment:
numpy concatenate of two arrays allocates new memory to create a new bigger array where the data from the two arrays together will be concatenated. You really want to avoid doing this if you just need to compute sums...

Reply:
I've been thinking about this and, at the moment, could not think of a way to optimize this. Because we are not just computing sums. We are computing sums but we need to respect the axis along which the sums must be computed.

For example, for the dataset:

```python
dataset = [
[1, 2],
[3, 4], 
[5, 6]
]
```

If we want to compute the mean along axis=0 then the cumulative sum is:

```python
# first iteration
[[1, 2]]
# second iteration
[[1, 2]] (+ on axis 0) (sum [[3, 4]] on axis 0) = [[1, 2]] (+ on axis 0) [[3, 4]] = [[4, 6]]
# forth iteration
[[9, 12]]
```

But for axis=1:

```python
# first iteration
[[3]]
# secodn iteration
[[3]] (+ on axis 1) (sum [[3, 4]] on axis 1) = [[3]] (+ on axis 1) [[7]] = [[3], [7]] # Involves concatenation either before or after summing
# third iteration
[[3], [7]] (+ on axis 1) (sum [[4, 6]] on axis 1) = [[3], [7]] (+ on axis 1) [[10]] = [[3], [7], [10]]
```

Which, for both cases a a particular case that, when receiving several samples at the same time the total sum is just `sum_so_far + (sum received samples)` but, for our case, with respect to a given axis which, may involve concatenation either before or after the sum due to the streaming nature of the problem.

I have also tested with `np.stack`, `np.vstack` but they all copy memory as well.
This copying is very small because it duplicates only temporarily one example of the dataset when we may already have thousands in memory.

We could perhaps drop entirely this `streaming_mean` across several dimensions at the same time and focus just on computing the mean for particular cases where we need them. And leave this general case for later if the need arises
For example, if we want to normalize the wind pressures, we can use a normal streaming mean and just iterate over each value of each example on the dataset.

Reply:
Thanks Augusto for the additional details. I didn't know you wanted to support that case of axis=1 which in fact the "accumulator" keeps growing in size (more elements in the array) --- that seems a little unexpected because the final size will be as long as the total stream size (collapsed along one dimension), so there is not much gain from computing online statistics... Do we need that?

Reply:
We do not in fact need the `axis=1`. I think we can infact go simpler. And just compute the mean of the predictions that we want.

We do need to have a way to do this in a streaming fashion to provide support for datasets in streaming mode

Reply:
I will remove the support for axis=1

Reply:
Great, thanks!